lib/util/make-knex.js

```diff
@@ -138,89 +229,46 @@ function _copyEventListeners(eventName, sourceKnex, targetKnex) {
 function redefineProperties(knex, client) {
   // Allow chaining methods from the root object, before
   // any other information is specified.
+  //
+  // TODO: `QueryBuilder.extend(..)` allows new QueryBuilder
+  //       methods to be introduced via external components.
+  //       As a side-effect, it also pushes the new method names
+  //       into the `QueryInterface` array.
+  //
+  //       The Problem: due to the way the code is currently
+  //       structured, these new methods cannot be retroactively
+  //       injected into existing `knex` instances!  As a result,
+  //       some `knex` instances will support the methods, and
+  //       others will not.
+  //
+  //       We should revisit this once we figure out the desired
+  //       behavior / usage.  For instance: do we really want to
+  //       allow external components to directly manipulate `knex`
+  //       data structures?  Or, should we come up w/ a different
+  //       approach that avoids side-effects / mutation?
+  //
+  //      (FYI: I noticed this issue because I attempted to integrate
+  //       this logic directly into the `KNEX_PROPERTY_DEFINITIONS`
+  //       construction.  However, `KNEX_PROPERTY_DEFINITIONS` is
+  //       constructed before any `knex` instances are created.
+  //       As a result, the method extensions were missing from all
+  //       `knex` instances.)
```

Comment:
I decided to omit this part from the PR until we have a chance to discuss it.

On one hand, we could probably work around this by using an actual `Proxy(..)`.  On the other hand, it seems strange that we're allowing `knex` data / structures to be directly manipulated in this way.  Perhaps we can come up w/ some way to preserve the original `QueryBuilder`, yet also allow end-users to create/use custom `QueryBuilder`s?